runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/status/ChannelzServlet.java

```diff
@@ -0,0 +1,245 @@
+package org.apache.beam.runners.dataflow.worker.status;
+
+import org.apache.beam.vendor.grpc.v1p60p1.io.grpc.channelz.v1.*;
+import org.apache.beam.vendor.grpc.v1p60p1.io.grpc.protobuf.services.ChannelzService;
+import org.apache.beam.vendor.grpc.v1p60p1.io.grpc.stub.StreamObserver;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.util.concurrent.SettableFuture;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/** Respond to /channelz with the GRPC channelz data. */
+@SuppressWarnings({
+        "nullness" // TODO(https://github.com/apache/beam/issues/20497)
+})
+public class ChannelzServlet extends BaseStatusServlet implements DebugCapture.Capturable {
+    private static final String PATH = "/channelz";
+    private static final Logger LOG = LoggerFactory.getLogger(ChannelzServlet.class);
+    private static final int MAX_TOP_CHANNELS_TO_RETURN = 100;
+
+    private final ChannelzService channelzService = ChannelzService.newInstance(MAX_TOP_CHANNELS_TO_RETURN);
+
+    public ChannelzServlet() {
+        super(PATH);
+    }
+
+    @Override
+    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
+        response.setStatus(HttpServletResponse.SC_OK);
+        PrintWriter writer = response.getWriter();
+        captureData(writer);
+    }
+
+    @Override
+    public String pageName() {
+        return PATH;
+    }
+
+    @Override
+    public void captureData(PrintWriter writer) {
+        writer.println("<html>");
+        writer.println("<h1>Channelz</h1>");
+        appendTopChannels(writer);
+        writer.println("</html>");
+    }
+
+    // channelz proto says there may not be cycles in the ref graph
```

Comment:
nit: period at end of comment
also a little confusing, does it say there may be cycles? or are we just being extra careful to prevent a possible cycle?

Reply:
done